### PR TITLE
Add support for Ceph RGW Object Store

### PR DESCRIPTION
**Pull request by wido**
### Description

This PR adds support for the Ceph RGW as an Object Store Driver in CloudStack

### Types of changes

- [x] New feature (non-breaking change which adds functionality)

### Screenshots (if appropriate):

Screenshot from 2023-12-20 17-10-04

### How Has This Been Tested?

I have a local Ceph environment (IPv6 only) running with the RADOS Gateway (S3 server) running against which I tested this PR.

It also has SSL enabled so it could verify that this works as expected.

#### How did you try to break this feature and the system with this change?

I manually tried to create and remove buckets and that worked as expected.

### CODE REVIEW

**Comment by rohityadavcloud** [comment]
@wido I'll help kick tests again - can you check for build failures?

@blueorangutan package

File: plugins/storage/object/ceph/pom.xml
Comment (weizhouapache) [inline plugins/storage/object/ceph/pom.xml:27]:
```suggestion
        <version>4.20.0.0-SNAPSHOT</version>
```

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
test failure found
```
Error:  Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.059 s <<< FAILURE! - in org.apache.cloudstack.storage.datastore.provider.CephObjectStoreProviderImplTest
Error:  testGetName(org.apache.cloudstack.storage.datastore.provider.CephObjectStoreProviderImplTest)  Time elapsed: 0.01 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<Ceph[ RGW]> but was:<Ceph[]>
	at org.apache.cloudstack.storage.datastore.provider.CephObjectStoreProviderImplTest.testGetName(CephObjectStoreProviderImplTest.java:41)
```

cc @wido

File: server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java
Comment (weizhouapache) [inline server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java:237]: with this change, template registration from IPv6 server will be possible
need testing

**Comment by weizhouapache** [comment]
@blueorangutan package

File: plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CephObjectStoreLifeCycleImpl.java
Comment (JoaoJandre) [inline plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CephObjectStoreLifeCycleImpl.java:41]:
```suggestion
    protected Logger logger = LogManager.getLogger(getClass());
```
Also need to update the logger import to log4j2

File: plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/driver/CephObjectStoreDriverImpl.java
Comment (JoaoJandre) [inline plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/driver/CephObjectStoreDriverImpl.java:64]: You don't need to create loggers for each new command anymore, they will inherit it from the BaseObjectStoreDriverImpl because of #7131

**Review by JoaoJandre — COMMENTED** [review]
Hey, @wido please merge this PR with main and change the loggers names/imports. https://github.com/apache/cloudstack/pull/7131's description has a script you can use if you want to (after merging with main).

File: plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CephObjectStoreLifeCycleImpl.java
Comment (wido) [reply plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CephObjectStoreLifeCycleImpl.java]: Done!

**Comment by wido** [comment]
I just rebased and implemented the new logger way of working.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
> Code wise this PR is now good? We are just waiting for the tests to proceed, correct?

@wido 
it is great.
I will setup a ceph environment and do some testing (probably next week).

**Comment by DaanHoogland** [comment]
@wido the builds fail with 
```
20:25:32 [ERROR] Errors: 
20:25:32 [ERROR]   UriUtilsTest.validateUrl:285 ? IllegalArgument Unable to resolve ipv6.google.c...
```
I don't understand the lint failure in the GHA, as the file reported does not seem to be touched in this PR.

**Comment by wido** [comment]
> > Code wise this PR is now good? We are just waiting for the tests to proceed, correct?
> 
> @wido it is great. I will setup a ceph environment and do some testing (probably next week).

Ok! Keep in mind that getting the object browser to work in the UI will be difficult. This is due to CORS and outside the scope of this PR.

That is something we need to fix regardless of this PR

**Comment by wido** [comment]
> @wido the builds fail with
> 
> ```
> 20:25:32 [ERROR] Errors: 
> 20:25:32 [ERROR]   UriUtilsTest.validateUrl:285 ? IllegalArgument Unable to resolve ipv6.google.c...
> ```
> 
> I don't understand the lint failure in the GHA, as the file reported does not seem to be touched in this PR.

I am touchting the UriUtils.java file and removing an IPv6 check there. I'll have to verify why this fails then

**Comment by DaanHoogland** [comment]
> I don't understand the lint failure in the GHA, as the file reported does not seem to be touched in this PR.

this was about the lint failure, but that passes now.

**Comment by wido** [comment]
Can I ping somebody to get this one merged?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@wido the title says draft, is this ready?

**Comment by wido** [comment]
> @wido the title says draft, is this ready?

Ah, just removed the Draft :-)

File: utils/src/test/java/com/cloud/utils/UriUtilsTest.java
Comment (shwstppr) [inline utils/src/test/java/com/cloud/utils/UriUtilsTest.java:286]: This testcase is failing during BO build. Do we need to add any guard for ipv6 support? cc @wido 

```
13:18:52 [INFO] Running com.cloud.utils.UriUtilsTest
13:18:52 [ERROR] Tests run: 15, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.139 s <<< FAILURE! - in com.cloud.utils.UriUtilsTest
13:18:52 [ERROR] validateUrl(com.cloud.utils.UriUtilsTest)  Time elapsed: 0.113 s  <<< ERROR!
13:18:52 java.lang.IllegalArgumentException: Unable to resolve ipv6.google.com
13:18:52 	at com.cloud.utils.UriUtilsTest.validateUrl(UriUtilsTest.java:285)
```

Comment (wido) [reply utils/src/test/java/com/cloud/utils/UriUtilsTest.java:286]: We need to make sure that this code can work with IPv6 as well. But the problem is that the host where the build is running on doesn't support IPv6 and thus fails to run the test properly. This also explains why it doesn't fail on my local machines as they do have IPv6.

Comment (rohityadavcloud) [reply utils/src/test/java/com/cloud/utils/UriUtilsTest.java:286]: @wido could you write the test such that it can skip if host does not have ipv6 enabled?

Comment (wido) [reply utils/src/test/java/com/cloud/utils/UriUtilsTest.java:286]: > @wido could you write the test such that it can skip if host does not have ipv6 enabled? But still work in env like Github Actions where ipv6 is available

That's something you can't detect. I think we would have to catch the IllegalArgumentException in this case

Comment (shwstppr) [reply utils/src/test/java/com/cloud/utils/UriUtilsTest.java:286]: Maybe by checking if any of the network interfaces on the system have an IPv6 address?

Comment (wido) [reply utils/src/test/java/com/cloud/utils/UriUtilsTest.java:286]: > Maybe by checking if any of the network interfaces on the system have an IPv6 address?

That wouldn't give you a real idea if IPv6 is available yes or now. In the end you can't know, you just have to try and see what happens.

Comment (piyushvijay) [reply utils/src/test/java/com/cloud/utils/UriUtilsTest.java]: Is it possible to enable ipv6 on the hosts where this test is run?

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@wido can you address build failure:
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project cloud-utils: Compilation failure
Error:  /home/runner/work/cloudstack/cloudstack/utils/src/main/java/com/cloud/utils/UriUtils.java:[320,54] error: cannot find symbol
Error:    symbol:   class Inet6Address
Error:    location: class UriUtils
Error:  -> [Help 1]
```

File: utils/src/main/java/com/cloud/utils/UriUtils.java
Comment (rohityadavcloud) [inline utils/src/main/java/com/cloud/utils/UriUtils.java:26]: @wido I think this needs to be removed - looks like you've turned off maintainers to be able to make edits.

**Comment by rohityadavcloud** [comment]
@blueorangutan package

File: plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/driver/CephObjectStoreDriverImpl.java
Comment (rohityadavcloud) [inline plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/driver/CephObjectStoreDriverImpl.java:49]:
```suggestion
import org.apache.logging.log4j.Logger;
```

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@wido :
```
11:03:41 [ERROR] /jenkins/workspace/acs-deb-pkg-builder/plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CephObjectStoreLifeCycleImpl.java:[30,23] error: package org.apache.log4j does not exist
11:03:41 [ERROR] /jenkins/workspace/acs-deb-pkg-builder/plugins/storage/object/ceph/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CephObjectStoreLifeCycleImpl.java:[41,25] error: cannot find symbol
11:03:41   symbol:   class Logger
11:03:41   location: class CephObjectStoreLifeCycleImpl
```
strange this error would only happen on the debian build and the el build is successful, but there you have it ???

> Packaging result [SF]: ✔️ el8 ✔️ el9 ✖️ debian ✔️ suse15. SL-JID 10989

**Comment by rohityadavcloud** [comment]
Fixed some issue, tested LGTM but some cases don't work will raise a separate issue for @wido to explore further:

- Bucket creation fails when any of the options are specified or these options aren't supported: quota, versioning & encryption, bucket policy
- During Bucket creation, locking feature/option is ignored
- User quota not enforced/configured
- Bucket policy fails and throws exception:
```
Error while updating bucket. Error while updating bucket: admin-test-bucket. SignatureDoesNotMatch (Service: Amazon S3; Status Code: 403; Error Code: SignatureDoesNotMatch; Request ID: tx00000c2e2357a96c47d4c-0066d97a79-42778b-default; S3 Extended Request ID: 42778b-default-default; Proxy: null)
```
- UI browser doesn't work (due to cors issue, couldn't figure out how to configure radosgw for that)

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Review by rohityadavcloud — APPROVED** [review]
LGTM, tested this plugins with Ceph Reef rgw, doesn't support all options but good to merge this.

**Comment by rohityadavcloud** [comment]
Testing notes:

1. Setup ceph cluster & rgw service:
Screenshot 2024-09-05 at 1 28 35 PM

2. Next, was able to add this as object storage pool in cloudstack using url & api/secret credentials by creating a system 'admin' user in the rgw:
Screenshot 2024-09-05 at 1 29 31 PM

3. Was able to create bucket without any options or bucket policy, but then was able to use S3 client like Cyberduck to create/configure access permissions on the bucket & upload/download objects/files in it:
Screenshot 2024-09-05 at 2 04 52 PM

4. Bucket policies don't work via cloudstack UI though and throw exception:

Screenshot 2024-09-05 at 3 01 42 PM

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
Merging this based on manual QA with limitations and followup logged at #9641 (cc @wido), reviews, earlier smoketests & GH build/smoketests.

**Pull request merged**
